src/math/trig.c

```diff
@@ -99,3 +100,88 @@ int32_t sin_fixed(int32_t th_rad_fxp)
+ *	| thRadFxp	   | cdccosth	     |thRadFxp|cdccosth|
+ *	+----+-----+-------+----+----+-------+--------+--------+
+ *	|WLen| FLen|Signbit|WLen|FLen|Signbit| Qformat| Qformat|
+ *	+----+-----+-------+----+----+-------+--------+--------+
+ *	| 32 | 28  |  1	   | 32	| 31 |	 1   | 4.28   | 1.31   |
+ *	+------------------+-----------------+--------+--------+
+ */
+int32_t cos_fixed(int32_t th_rad_fxp)
+{
+	/*cordic_atan2_lookup_table = atan(2.^-(0:N-1)) N = 31 */
+	/*CORDIC Gain is cordic_gain = prod(sqrt(1 + 2.^(-2*(0:31-1)))) */
+	/*Inverse CORDIC Gain,inverse_cordic_gain = 1 / cordic_gain */
+	static const int32_t inpLUT[CORDICCOSINE_TABLE_SIZE] = { 843314857, 497837829,
+	263043837, 133525159, 67021687, 33543516, 16775851, 8388437, 4194283, 2097149,
+	1048576, 524288, 262144, 131072, 65536, 32768, 16384, 8192, 4096, 2048, 1024,
+	512, 256, 128, 64, 32, 16, 8, 4, 2, 1 };
+	int32_t b_idx;
+	int32_t b_yn;
+	int32_t xn;
+	int32_t xtmp;
+	int32_t ytmp;
+	int sign = 1;
+	/* 652032874 , deg = 69.586061*/
+	const int32_t CORDICCOSINE_LUT_Q29FL  =	 Q_CONVERT_FLOAT(1.214505869895220, 29);
+	/* 421657428 , deg = 90.000000 */
+	const int32_t CORDICCOSINE_PIOVERTWO_Q28FL  = Q_CONVERT_FLOAT(M_PI / 2, 28);
+	/* 843314857,  deg = 90.000000	*/
+	const int32_t CORDICCOSINE_PIOVERTWO_Q29FL  = Q_CONVERT_FLOAT(M_PI / 2, 29);
+	/* 1686629713, deg = 90.000000	*/
+	const int32_t CORDICCOSINE_PIOVERTWO_Q30FL  = Q_CONVERT_FLOAT(M_PI / 2, 30);
+
+	/*Addition or subtraction by a multiple of pi/2 is done in the data type*/
+	/*of the input. When the fraction length is 29, then the quantization error*/
+	/*introduced by the addition or subtraction of pi/2 is done with 29 bits of*/
+	/*precision.Input range of cordicsin must be in the range [-2*pi, 2*pi),*/
+	/*a signed type with fractionLength = wordLength-4 will fit this range*/
+	/*without overflow.Increase of fractionLength makes the addition or */
+	/*subtraction of a multiple of pi/2 more precise */
+
+	if (th_rad_fxp > CORDICCOSINE_PIOVERTWO_Q28FL) {
+		if ((th_rad_fxp - CORDICCOSINE_PIOVERTWO_Q29FL) <= CORDICCOSINE_PIOVERTWO_Q28FL) {
+			th_rad_fxp -= CORDICCOSINE_PIOVERTWO_Q29FL;
+			sign  = -1;
+		} else {
+			th_rad_fxp -= CORDICCOSINE_PIOVERTWO_Q30FL;
+
+		}
+	} else if (th_rad_fxp < -CORDICCOSINE_PIOVERTWO_Q28FL) {
+		if ((th_rad_fxp + CORDICCOSINE_PIOVERTWO_Q29FL) >= -CORDICCOSINE_PIOVERTWO_Q28FL) {
+			th_rad_fxp += CORDICCOSINE_PIOVERTWO_Q29FL;
+			sign  = -1;
+		} else {
+			th_rad_fxp += CORDICCOSINE_PIOVERTWO_Q30FL;
+		}
+	}
+	th_rad_fxp <<= 2;
+	b_yn = 0;
+	xn = CORDICCOSINE_LUT_Q29FL;
+	xtmp = CORDICCOSINE_LUT_Q29FL;
+	ytmp = 0;
+
+	for (b_idx = 0; b_idx < CORDICCOSINE_TABLE_SIZE; b_idx++) {
+		if (th_rad_fxp < 0) {
+			th_rad_fxp += inpLUT[b_idx];
+			xn += ytmp;
+			b_yn -= xtmp;
+		} else {
+			th_rad_fxp -= inpLUT[b_idx];
+			xn -= ytmp;
+			b_yn += xtmp;
+		}
+	xtmp = xn >> (b_idx + 1);
+	ytmp = b_yn >> (b_idx + 1);
+	}
+	th_rad_fxp = sign * xn;
```

Comment:
Could you make a common inline function to calculate xn and b_yn and handle with sin() and cos() the minor difference? Then also there would be need to optimize (intrinsics) only a single function.

Reply:
@singalsu , That's correct. sine and cosine are separated by xn and b_yn . I will make common inline function handler to make the difference.